File: syncfs.pl
#!/usr/bin/perl
package SyncFSFile;

use Data::Dumper;

sub bucket {
    my ($n) = shift;
    if ($n == 0) {
	return 0;
    }
    if ($n < 0) {
	return - bucket(-$n);
    }
    my $bucket = 0;
    while ($n) {
	$bucket++;
	$n = int($n/2);
    }
    return $bucket;
}

sub status_time_stats {
    my $time0 = time;
    for my $status (sort { $a cmp $b } keys %SyncFSFile::by_status_time) {
	my $h = $SyncFSFile::by_status_time{$status};
	my %buckets;
	for my $time (sort { $a <=> $b } keys %$h) {
	    my $count = 0;
	    my $h = $h->{$time};
	    for my $file (values %$h) {
		$buckets{bucket($time0 - $time)}++;
	    }
	}
	for my $bucket (sort { $a <=> $b } keys %buckets) {
	    print "$status $bucket " . $buckets{$bucket} . "\n";
	}
    }
}

sub new {
    my $class = shift;
    my $path = shift;
    my $time = time;

    return bless {
	path => $path,
	btime => $time,
	old_by_time => $time,
	status => "unknown"
    }, $class;
}

sub status {
    my ($self, $newstatus) = @_;
    return $self->{status} unless defined $newstatus;

    my $status = $self->{status};
    my $time = $self->{old_by_time};
    delete $SyncFSFile::by_status_time{$status}{$time}{$self};
    if (scalar keys %{$SyncFSFile::by_status_time{$status}{$time}} == 0) {
	delete $SyncFSFile::by_status_time{$status}{$time};
    }

    $self->{status} = $newstatus;

    $SyncFSFile::by_status_time{$newstatus}{$time}{$self} = $self;

    return $newstatus;
}

sub old_by_time {
    my ($self, $newtime) = @_;
    return $self->{old_by_time} unless defined $newtime;

    my $status = $self->{status};
    my $time = $self->{old_by_time};
    delete $SyncFSFile::by_status_time{$status}{$time}{$self};
    if (scalar keys %{$SyncFSFile::by_status_time{$status}{$time}} == 0) {
	delete $SyncFSFile::by_status_time{$status}{$time};
    }

    $self->{old_by_time} = $newtime;

    $SyncFSFile::by_status_time{$status}{$newtime}{$self} = $self;

    return $newtime;
}

sub touch {
    my ($self, $h, $done, $delay) = @_;
    $self->{time} = time;
    if ($h->{command} eq "write" and !$done) {
	$self->{modbytes} += $h->{size} + length $h->{path};
	$self->{edited_by}->{$h->{user}}++;
	my $cmdline = $h->{cmdline}->[0];
	$self->{cmdlines}->{$cmdline} = $cmdline;
    } elsif (($h->{command} eq "write") and $done) {
	$self->status("written");
	$self->old_by_time(time + $delay);
    } elsif ($h->{command} eq "symlink" and !$done) {
	$self->{modbytes} += $h->{size} + length $h->{path};
	$self->{edited_by}->{$h->{user}}++;
	my $cmdline = $h->{cmdline}->[0];
	$self->{cmdlines}->{$cmdline} = $cmdline;
    } elsif (($h->{command} eq "symlink") and $done) {
	$self->status("written");
	$self->old_by_time(time + $delay);
    } elsif (($h->{command} eq "create") and $done) {
	$self->status("written");
	$self->old_by_time(time + $delay);
    } elsif (($h->{command} eq "unlink") and $done) {
	$self->status("deleted");
	$self->old_by_time(time + $delay);
    } elsif (($h->{command} eq "unlink") and !$done) {
	$self->status("unknown");
    } elsif (($h->{command} eq "rename") and !$done) {
	$self->status("unknown");
    }
}

sub touch_renamed_to {
    my ($self, $h, $done) = @_;
    $self->{time} = time;
    if (!$done) {
	$self->status("unknown");
    } else {
	$self->status("written");
    }
}

sub touch_renamed_from {
    my ($self, $h, $done) = @_;
    $self->{time} = time;
    if (!$done) {
	$self->status("unknown");
    } else {
	$self->status("deleted");
    }
}

sub score {
    my ($self) = @_;
    my $age = time - $self->{time};
    my $modbytes = (1 + $self->{modbytes});
    my $write_time = $self->{time} - $self->{btime};

    return ($age + 1) * $modbytes;
}

sub path {
    my $self = shift;
    return $self->{path};
}

sub message {
    my $self = shift;
    $self->{message} .= "automatic commit\n\n";
    for my $cmdline (sort %{$self->{cmdlines}}) {
	$self->{message} .= $self->path . " " . "$cmdline\n";
    }
    $self->{cmdlines} = {};
    my $message = $self->{message};
    $self->{message} = "";
    return $message;
}

sub sync {
    my $self = shift;
    $self->{modbytes} = 0;
    $self->status("synched");
    $self->old_by_time(time + 120);
}

package main;

use strict;
use Mojo::JSON;
use File::Slurp qw(read_file);
use Data::Dumper;
use User::pwent;
use IPC::Run qw(run start);
use AnyEvent;
use AnyEvent::Handle;

my $timer_needed = 0;
my $fifos = shift;

my %files;

sub sync {
    my $stdin = join("\0", map { $_->path } sort { $b <=> $a } values %files);
    my $stdout = "";
    my $stderr = "";
    run(["echo", "rsync", "--update", "--files-from=-", "--from0", "mount", "pip\@10.4.0.1:syncfs/"], \$stdin, \$stdout, \$stderr);
    warn "stdout $stdout";
    warn "stderr $stderr";
}

sub del_files {
    my $stdin = "";
    my $maxfiles = 1024;
    my $i = 0;
    my @files;
    my $message = "";
    for my $file (sort { $b->score <=> $a->score } values %files) {
	next if $file->status ne "deleted";
	if ($i++ == $maxfiles) {
	    $timer_needed = 1;
	    last;
	}
	push @files, $file;
	$message .= $file->message;
    }

    if (@files) {
	my $stdin = join("\0", map { $_->path } @files) . "\0";
	eval {
	    my $succ = run(["git", "rm", "--ignore-unmatch", "--pathspec-from-file=-", "--pathspec-file-nul"], \$stdin);
	    if ($succ) {
		$succ = run(["git", "commit", "--allow-empty", "-m", substr($message, 0, 1024)]);
	    }
	    for my $file (@files) {
		$file->sync;
	    }
	};

	warn $@ if $@;

	return 1;
    }

    return 0;
}

sub add_files {
    my %opts = @_;
    my $stdin = "";
    my $maxfiles = 1024;
    my $i = 0;
    my @files;
    my $message = "";
    for my $file (sort { $b->score <=> $a->score } values %files) {
	next if $file->status ne "written";
	next if $opts{nolarge} and $file->{modbytes} > 1024 * 1024;
	last if $file->score == 0;
	if ($i++ == $maxfiles) {
	    $timer_needed = 1;
	    last;
	}
	push @files, $file;
	$message .= $file->message;
    }

    if (@files) {
	my $stdin = join("\0", map { $_->path } @files) . "\0";
	my $succ = run(["git", "add", "--ignore-removal", "--pathspec-from-file=-", "--pathspec-file-nul"], \$stdin);
	if ($succ) {
	    $succ = run(["git", "commit", "--allow-empty", "-m", $message]);
	}
	my $rev = "";
	if ($succ) {
	    $stdin = "";
	    $succ = run(["git", "rev-parse", "HEAD"], \$stdin, \$rev);
	    chomp $rev;
	}
	if (!$succ) {
	    die;
	}
	for my $file (@files) {
	    $file->sync;
	}

	return 1;
    }

    return 0;
}

my @lowerer_queue;
my $lowerer_handle;
my $lowerer_fh;

sub lower_files {
    return 0; # disabled; overlayfs is buggy
    my %opts = @_;
    my $maxfiles = 32;
    my @files;
    my $i = 0;
    for my $file (sort { $b->score <=> $a->score } values %files) {
	next if $file->status ne "synched";
	if ($i++ == $maxfiles) {
	    $timer_needed = 1;
	    last;
	}
	push @files, $file;
    }

    if (@files) {
	for my $file (@files) {
	    if (@lowerer_queue == 0) {
		warn "lowerer_queue now " . scalar(@lowerer_queue);
		push @lowerer_queue, $file;
		$lowerer_handle = new AnyEvent::Handle(
		    fh => $lowerer_fh,);
		$lowerer_handle->push_write($file->path . "\n");
	    } else {
		push @lowerer_queue, $file;
	    }
	}

	return 1;
    }

    return 0;
}

my $resolve_starid_last_pid;
my $resolve_starid_user;
my $resolve_starid_cmdline;
sub resolve_starid {
    my ($h) = @_;
    if ($h->{pid} eq $resolve_starid_last_pid) {
	$h->{cmdline} = $resolve_starid_cmdline;
	$h->{user} = $resolve_starid_user;
    }
    $resolve_starid_last_pid = $h->{pid};
    eval {
	my @cmdline = split("\0", read_file("/proc/" . $h->{pid} . "/cmdline"));
	$resolve_starid_cmdline = $h->{cmdline} = \@cmdline;
	$resolve_starid_user = $h->{user} = getpwuid($h->{uid})->name;
    };
}

my %editors = (
    "emacs",
    "vi",
    "vim",
    );

sub update_score {
    my ($h, $done) = @_;
    if (!exists $files{$h->{path}}) {
	$files{$h->{path}} = new SyncFSFile($h->{path});
    }
    my $file = $files{$h->{path}};
    my $delay = 5;
    for my $cmdline (values %{$file->{cmdlines}}) {
	for my $cmd (split "/", $cmdline) {
	    $delay = 0 if $editors{$cmd};
	}
    }
    $file->touch($h, $done, $delay);
    return $delay;
}

sub update_rename_score {
    my ($h, $done) = @_;
    if (!exists $files{$h->{path1}}) {
	$files{$h->{path1}} = new SyncFSFile($h->{path1});
    }
    if (!exists $files{$h->{path2}}) {
	$files{$h->{path2}} = new SyncFSFile($h->{path2});
    }
    {
	my $file = $files{$h->{path1}};
	$file->touch_renamed_from($h, $done);
    }
    {
	my $file = $files{$h->{path2}};
	$file->touch_renamed_to($h, $done);
    }

    return 5;
}

sub contact_sync_host {
    my ($host) = @_;
    my $stdout = "";
    my $stderr = "";
    my $stdin = "";
    run(["ssh", "$host", "echo", "foo", ">>", "sync/syncfs-pings"], \$stdin, \$stdout, \$stderr);
}

my $synctime = 0;

my $fh;
open $fh, "$fifos/fuse-to-daemon" or die;
my $outfh;
open $outfh, ">$fifos/daemon-to-fuse" or die;
my $notifyfh;
open $notifyfh, ">$fifos/daemon-to-notify" or die;
open $lowerer_fh, ">$fifos/daemon-to-lowerer" or die;
chdir "c00git";

my $timer_last_run = 0;
my $timer_running = 0;

sub run_timer {
    my %opts = @_;
    # SyncFSFile::status_time_stats;
    $timer_last_run = time;
    return if ($timer_running);
    $timer_running = 1;
    eval {
	my $rev;
	if (($rev = add_files(%opts)) ||
	    ($rev = del_files(%opts))
	    # || ($rev = lower_files(%opts))
	    ) {
	    print $notifyfh $rev . "\n";
	    flush $notifyfh;
	}
    };
    $timer_running = 0;
    $timer_last_run = time;
}

my $timer_last_started = 0;
my $timer;
sub check_timer;
sub check_timer {
    my $time = time;
    if ($time >= $timer_last_started + 5) {
	$timer_needed = 0;
	$timer_last_started = $time;
	run_timer;
    } elsif ($time - $timer_last_run > 60) {
	$timer_needed = 0;
	$timer_last_started = $time;
	run_timer(nolarge => 1);
    }
    $timer = AE::timer 5, 0, sub { check_timer };
    $timer_last_started = $time;
}

my $cv = AnyEvent->condvar();
my $hash;
my $hdl; $hdl = new AnyEvent::Handle(
    fh => $fh,
    on_read => sub {
	shift->unshift_read(line => sub {
	    my ($h, $line) = @_;
	    my $delay;
	    warn "$line\n" if $line ne "";
	    if ($line ne "") {
		$hash = Mojo::JSON::decode_json($line);
		resolve_starid($hash);
		$outfh->print("\n");
		$outfh->flush;
		if ($hash->{command} eq "rename") {
		    $delay = update_rename_score($hash, 0);
		} else {
		    $delay = update_score($hash, 0);
		}
	    } else {
		$outfh->print("\n");
		$outfh->flush;
		if ($hash->{command} eq "rename") {
		    $delay = update_rename_score($hash, 1);
		} else {
		    $delay = update_score($hash, 1);
		}
	    }

	    if ($delay == 0) {
		run_timer(nolarge => 1);
		$timer = AE::timer 5, 0, sub {
		    check_timer;
		};
		$timer_last_started = time;
	    } else {
		check_timer;
		$timer = AE::timer 5, 0, sub {
		    check_timer;
		};
		$timer_last_started = time;
	    }
			    });
    },
    on_eof => sub { $cv->send },
    on_error => sub { $cv->send });

warn "starting";
$cv->recv();
warn "exiting";

File: syncfs.mjs
import * as oldfs from "fs";
import * as fs from "fs/promises";
import * as readline from "readline";
import * as child_process from "child_process";
import { setTimeout, clearTimeout } from "timers";
import * as syncfs from "./syncfs-lib.mjs";

let [fifopath, ...remotes] = process.argv.slice(2);

function SyncFSFile(path, event)
{
    this.path = path;
    this.time = Date.now() / 1000.0;
    this.old_by_time = this.time;
    this.set_state("unknown");
    this.factor = 1.0;
    if (path.match(/tmp/))
	this.factor *= .1;
    if (path.match(/lock/))
	this.factor *= .1;
    if (path.match(/#/))
	this.factor *= .1;
    if (path.match(/~$/))
	this.factor *= .1;
    if (event.cmdline[0].match(/emacs/))
	this.factor *= 10;
    this.score = 0.0;
    this.delta_size = 0;
}

SyncFSFile.map_by_path = new Map();
SyncFSFile.by_path = function (path, event) {
    if (!SyncFSFile.map_by_path.has(path))
	SyncFSFile.map_by_path.set(path, new SyncFSFile(path, event));
    return SyncFSFile.map_by_path.get(path);
};

SyncFSFile.map_by_state = new Map();
SyncFSFile.map_by_state.set(undefined, new Set());
SyncFSFile.map_by_state.set("written", new Set());
SyncFSFile.map_by_state.set("deleted", new Set());
SyncFSFile.map_by_state.set("unknown", new Set());
SyncFSFile.map_by_state.set("synched", new Set());

SyncFSFile.prototype.map_by_state = SyncFSFile.map_by_state;

SyncFSFile.by_state = async function (state, n, _time)
{
    let files = [];
    let time = Date.now()/1000.0;
    for (let file of this.map_by_state.get(state)) {
	if (file.old_by_time < time)
	    files.push(file);
    }
    files.map(file => file.update_score(time));
    files = files.sort((a, b) => b.score - a.score);
    if (state === "written")
	for (let i = 0; i < files.length; i++)
	    try {
		(await fs.open(files[i].path)).close();
	    } catch (e) {
		files.splice(i, 1);
		i--;
	    }
    console.log(`${files.length} files in state ${state}`)
    files.splice(n);
    for (let file of files) {
	//console.log(file.path, file.score);
    }
    return files;
};

SyncFSFile.prototype.update_score = function (time = Date.now() / 1000.0)
{
    let dt = time - this.time;
    //this.score *= Math.exp(dt * .01);
    this.time = time;
};

SyncFSFile.prototype.bump_score = function (event, i)
{
    this.update_score();
    this.delta_size += this.path.length + event.size;
    let absscore = this.path.length + event.size;
    let relscore = 1 - Math.exp(-absscore / 10000);
    this.score += relscore;
};

SyncFSFile.prototype.set_state = function (newstate)
{
    this.map_by_state.get(this.state).delete(this);
    this.state = newstate;
    this.map_by_state.get(this.state).add(this);
};

SyncFSFile.prototype.sync = function (rev)
{
    this.delta_size = 0;
    this.set_state("synched");
    this.score = 0;
};

SyncFSFile.prototype.set_old_by_time = function (old_by_time)
{
    this.old_by_time = old_by_time;
};

SyncFSFile.prototype.map_access_types = new Map();

SyncFSFile.prototype.map_access_types.set("very hot", { delay: 0, factor: 1000 });
SyncFSFile.prototype.map_access_types.set("medium", { delay: 15, factor: 1 });
SyncFSFile.prototype.map_access_types.set("cold", { delay: 120, factor: .1 });
SyncFSFile.prototype.map_access_types.set("very cold", { delay: 3600, factor: .01 });

SyncFSFile.prototype.determine_access_type = function (event, i)
{
    let path = this.path;
    let components = path.split("/");
    let is = {};
    if (event.cmdline) {
	let [command, ...args] = event.cmdline;
	for (let component of command.split("/")) {
	    if (component === "emacs")
		is.emacs = true;
	    if (component === "gcc" ||
		component === "g++" ||
		component === "cc1")
		is.compiled = true;
	}
    }
    for (let component of components) {
	if (component === ",git") {
	    is.git = true;
	}
	if (component.match(/^tmp/)) {
	    is.tmp = true;
	}
	if (component.match(/\.lock$/)) {
	    is.lock = true;
	}
	if (component.match(/\.pack$/)) {
	    is.pack = true;
	}
	if (component === "index") {
	    is.index = true;
	}
	if (component.match(/\.(c|h|s|cc|C|S|js|mjs|txt|sh|bash|mk)$/)) {
	    is.source = true;
	}
	if (component.match(/^(Makefile|GNU[mM]akefile)/)) {
	    is.source = true;
	}
	if (component.match(/\.(o|so|a|exe)$/)) {
	    is.compiled = true;
	}
    }
    if (is.source || is.emacs)
	return "very hot";
    if (is.git && (is.tmp || is.lock || is.index))
	return "very cold";
    if (is.git || is.tmp || is.lock)
	return "cold";
    if (is.compiled)
	return "cold";
    return "medium";
};

SyncFSFile.prototype.touch_written = function (event, i)
{
    this.set_state("written");
    this.bump_score(event, i);
    let at = this.map_access_types.get(this.determine_access_type(event, i));
    this.set_old_by_time(Date.now()/1000.0 + at.delay);
};

SyncFSFile.prototype.touch_unknown = function (event, i)
{
    this.set_state("unknown");
};

SyncFSFile.prototype.touch_deleted = function (event, i)
{
    this.set_state("deleted");
    this.bump_score(event, i);
    let at = this.map_access_types.get(this.determine_access_type(event, i));
    this.set_old_by_time(Date.now()/1000.0 + at.delay);
};

SyncFSFile.prototype.touch = function (event, i)
{
    if (event.done) {
	if (event.command === "create" ||
	    event.command === "write" ||
	    event.command === "symlink" ||
	    event.command === "rename" && i == 1)
	    this.touch_written(event, i);
	else if (event.command === "unlink" ||
		 event.command === "rename" && i == 0)
	    this.touch_deleted(event, i);
    } else {
	this.touch_unknown(event, i);
    }
}

async function update_score(event)
{
    let paths = [];
    if (event.path !== undefined)
	paths.push(event.path);
    else {
	paths.push(event.path1);
	paths.push(event.path2);
    }

    for (let i = 0; i < paths.length; i++) {
	let f = SyncFSFile.by_path(paths[i], event);
	f.touch(event, i);
    }
}

async function resolve_starid(event)
{
    try {
	let cmdline =
	    (await fs.readFile("/proc/" + event.pid + "/cmdline", "utf-8"))
	    .split("\0");
	event.cmdline = cmdline;
    } catch (e) {
    }
}

async function main()
{
    process.chdir("c00git");
    let input = oldfs.createReadStream("../" + fifopath + "/fuse-to-daemon");
    let output = oldfs.createWriteStream("../" + fifopath + "/daemon-to-fuse",
					 { highWaterMark: 0 });

    let rl = readline.createInterface({
	input,
    });

    let remote_resolve;
    let remote_async = async function () {
	while (true) {
	    let remote_promise = new Promise(r => remote_resolve = r);
	    let [rev, size] = await remote_promise;
	    await new Promise(r => {
		child_process.spawn("git", ["pull", remotes[0], rev, "--no-edit"], {stdio: ["inherit", "inherit", "inherit"]}).on("close", r);
	    });
	    await new Promise(r => {
		child_process.spawn("git", ["commit", "-m", "merge"], {stdio: ["inherit", "inherit", "inherit"]}).on("close", r);
	    });
	}
    };
    remote_async();

    let remote_setup; remote_setup = () => {
	let remote_input = oldfs.createReadStream("../" + fifopath + "/remote-to-local");
	let remote_rl = readline.createInterface({
	    input: remote_input,
	});
	let remote_q = [];

	remote_rl.on("line", async function (line) {
	    console.error("received " + line);
	    let [rev,size] = line.split(/ /);
	    remote_q.push([rev, size]);
	    if (remote_q.length > 1)
		return;
	    while (remote_q.length) {
		while (!remote_resolve)
		    await new Promise(r => setTimeout(r, 1000));
		remote_resolve(remote_q.shift());
		remote_resolve = undefined;
	    }
	});
	remote_rl.on("close", () => {
	    remote_setup();
	})
    };
    remote_setup();

    let notify_queues = new Map();

    async function notify(rev)
    {
	for (let remote of remotes) {
	    if (!notify_queues.has(remote))
		notify_queues.set(remote, []);
	    let q = notify_queues.get(remote);
	    q.push(rev);
	    if (q.length === 1) {
		let cb; cb = () => {
		    if (q.length) {
			child_process.spawn("ssh", [remote, "echo", q.shift(), ">>", "sync/fifos/remote-to-local"], {stdio: ["pipe", "inherit", "inherit"]}).on("close", cb);
		    }
		};
		cb();
	    }
	}
    }

    async function add_or_del_files(state, args)
    {
	let time = Date.now() / 1000.0;
	let max_files = 16 * 1024;
	let delta = 0;
	let files = await SyncFSFile.by_state(state, max_files, time);
	if (files.length === 0)
	    return;

	for (let file of files)
	    delta += file.delta_size;

	let paths = files.map(f => f.path);
	let stdin = paths.join("\0");
	let error = await new Promise(r => {
	    let child =
		child_process.spawn("git", args, {stdio: ["pipe", "inherit", "inherit"]});
	    child.on("close", code => r(code !== 0));
	    child.stdin.write(stdin);
	    child.stdin.end();
	});
	if (error)
	    throw error;
	error = await new Promise(r => {
	    let child =
		child_process.spawn("git", ["commit", "--allow-empty", "-m", "automatic commit"], {stdio: ["inherit", "inherit", "inherit"]})
	    child.on("close", code => r(code !== 0));
	});
	if (error)
	    throw error;
	let rev;
	error = await new Promise(r => {
	    let child =
		child_process.spawn("git", ["rev-parse", "HEAD"],
				    { stdio: ["inherit", "pipe", "inherit"] });
	    rev = "";
	    child.stdout.on("data", data => {
		rev += data.toString();
	    });
	    child.on("close", code => r(code !== 0));
	})
	if (error)
	    throw error;
	for (let file of files)
	    file.sync(rev);
	while (rev.length && rev[rev.length-1] === "\n")
	    rev = rev.substr(0, rev.length - 1);
	return rev + " " + delta;
    }

    async function add_files()
    {
	return await add_or_del_files("written", ["add", "--pathspec-from-file=-", "--pathspec-file-nul"]);
    }

    async function del_files()
    {
	return await add_or_del_files("deleted", ["rm", "--ignore-unmatch", "--pathspec-from-file=-", "--pathspec-file-nul"]);
    }

    async function handle_event(event)
    {
	await resolve_starid(event);
	output.write("\n");
	await update_score(event);
    }

    let event = {};
    let timer;
    let timerTriggered;
    let timerRunning;

    async function timeout()
    {
	timer = undefined;
	timerTriggered = true;
	if (timerRunning) {
	    return;
	}
	while (timerTriggered) {
	    timerRunning = true;
	    try {
		let rev;
		if ((rev = await add_files()) === undefined &&
		    (rev = await del_files()) === undefined) {
		    timerTriggered = false;
		} else {
		    notify(rev);
		}
	    } catch (e) {
	    }
	    timerRunning = false;
	    await new Promise(r => setTimeout(r, 1000));
	}
    }

    rl.on("line", async function (line) {
	//console.log(line);
	if (line === "") {
	    event.done = true;
	    await handle_event(event);
	} else {
	    event = JSON.parse(line);
	    event.done = false;
	    await handle_event(event);
	}
	if (timer !== undefined) {
	    clearTimeout(timer);
	}
	timer = setTimeout(timeout, 5000);
    });

    rl.on("close", function () {
	console.error("should be terminating properly");
    });
}

main();
